### PR TITLE
Speed up native assets target

### PR DESCRIPTION
**Pull request by dcharkes**
Speeds up the native assets target in the backend by

1. changing other targets `gen_dart_plugin_registrant` and `release_unpack_ios` to do async I/O,
2. not reparsing the package config, and
3. not calling `dart pub deps --json` for 0 or 1 packages (fixed package:native_assets_builder).

* https://github.com/flutter/flutter/issues/134427

```
           [   +2 ms] native_assets: Starting due to {}
           [   +2 ms] Skipping target: gen_localizations
           [   +1 ms] gen_dart_plugin_registrant: Starting due to {InvalidatedReasonKind.inputChanged: The following inputs have updated contents: /Users/dacoharkes/flt/engine/flutter/examples/hello_world/.dart_tool/package_config_subset}
           [  +33 ms] gen_dart_plugin_registrant: Complete
           [ +107 ms] release_unpack_ios: Complete
           [  +60 ms] Writing native_assets.yaml.
           [   +7 ms] Writing /Users/dacoharkes/flt/engine/flutter/examples/hello_world/.dart_tool/flutter_build/be2692bbfbc0b9a27fcd2422d52354c6/native_assets.yaml done.
           [        ] native_assets: Complete
```

->

```
           [   +4 ms] native_assets: Starting due to {}
           [        ] Skipping target: gen_localizations
           [   +1 ms] gen_dart_plugin_registrant: Starting due to {InvalidatedReasonKind.inputChanged: The following inputs have updated contents: /Users/dacoharkes/flt/engine/flutter/examples/hello_world/.dart_tool/package_config_subset}
           [  +31 ms] Writing native_assets.yaml.
           [   +8 ms] Writing /Users/dacoharkes/flt/engine/flutter/examples/hello_world/.dart_tool/flutter_build/f9451a65a465bfab70d004e21d6cc1d6/native_assets.yaml done.
           [   +1 ms] native_assets: Complete
```

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/lib/src/dart/package_map.dart [hunk 1]
```diff
@@ -25,10 +25,10 @@ Future<PackageConfig> loadPackageConfigWithLogging(File file, {
     file.absolute.uri,
     loader: (Uri uri) async {
       final File configFile = fileSystem.file(uri);
-      if (!configFile.existsSync()) {
+      if (!await configFile.exists()) {
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/dart/package_map.dart:28]: is it worth it to make this async?

File: packages/flutter_tools/lib/src/dart/package_map.dart [hunk 2]
```diff
@@ -25,10 +25,10 @@ Future<PackageConfig> loadPackageConfigWithLogging(File file, {
...
         return null;
       }
-      return Future<Uint8List>.value(configFile.readAsBytesSync());
+      return Future<Uint8List>.value(await configFile.readAsBytes());
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/dart/package_map.dart:31]: We don't need an await here, right?

```suggestion
      return configFile.readAsBytes();
```

File: packages/flutter_tools/lib/src/flutter_plugins.dart [hunk 3]
```diff
@@ -26,23 +26,29 @@ import 'platform_plugins.dart';
...
+  await file.writeAsString(renderedTemplate);
 }
 
-Plugin? _pluginFromPackage(String name, Uri packageRoot, Set<String> appDependencies, {FileSystem? fileSystem}) {
+Future<Plugin?> _pluginFromPackage(String name, Uri packageRoot, Set<String> appDependencies,
+    {FileSystem? fileSystem}) async {
   final FileSystem fs = fileSystem ?? globals.fs;
   final File pubspecFile = fs.file(packageRoot.resolve('pubspec.yaml'));
-  if (!pubspecFile.existsSync()) {
+  if (!await pubspecFile.exists()) {
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/flutter_plugins.dart:45]: same question

File: packages/flutter_tools/lib/src/native_assets.dart [hunk 4]
```diff
@@ -5,6 +5,7 @@
 // Logic for native assets shared between all host OSes.
 
 import 'package:logging/logging.dart' as logging;
+import 'package:native_assets_builder/native_assets_builder.dart';
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/native_assets.dart:8]: why are we importing the same library twice?

Comment (dcharkes) [reply packages/flutter_tools/lib/src/native_assets.dart:8]: We have conflicting types in the library and this file, so we need an import with a name for qualified access.
However, using all types from that library with a qualified name makes things very verbose.
Any suggestions?

File: packages/flutter_tools/lib/src/dart/package_map.dart (hunk 2 above)
Comment (dcharkes) [reply packages/flutter_tools/lib/src/dart/package_map.dart]: I had to revert this, for details see https://github.com/flutter/flutter/issues/134621.

File: packages/flutter_tools/lib/src/native_assets.dart (hunk 4 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/native_assets.dart:8]: Ahh, I see. could you hide the ambiguous identifier from the unqualified import, to make it more clear what's going on?

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/native_assets.dart:8]: or I guess, hide the import from the other library and then only use the unqualified import?

**Review by christopherfujino — APPROVED** [review]
LGTM

**Pull request merged**